xds/src/test/java/io/grpc/xds/LocalityStoreTest.java

```diff
@@ -171,8 +164,7 @@ public LoadBalancer newLoadBalancer(Helper helper) {
   @Before
   public void setUp() {
     doReturn(mock(ChannelLogger.class)).when(helper).getChannelLogger();
-    doReturn(mock(Subchannel.class)).when(helper).createSubchannel(
-        ArgumentMatchers.<EquivalentAddressGroup>anyList(), any(Attributes.class));
+    doReturn(mock(Subchannel.class)).when(helper).createSubchannel(any(CreateSubchannelArgs.class));
```

Comment:
One nit pick, but don't change it here as there are many other places have the same problem.

According to go/java-testing/mocking:

> Tip: Use Mockito.doReturn(value).when(mockObject).method(arguments) in cases where when(...) does not work. Avoid it otherwise (style rule, it is harder to read).

There is an external (non-google) version:
> The first way (when/then) is considered preferred because it is typesafe and because it’s more readable.

Reply:
I don't quite agree with the tip though.

`when().thenReturn()` is terribly unsafe. 
The following test looks pretty readable but fails.
```
  @Test
  public void dummy() {
    Object mock = mock(Object.class);
    when(mock.hashCode()).thenReturn(8);
    
    mock.hashCode();
    
    verify(mock, times(1)).hashCode(); // it is actually called twice
  }
```

I don't like to write `when().thenReturn()` at the beginning, some days later I add new tests and they fail, and I spend an hour of debugging to find out why: I have to modify it to `doReturn().when().method()`. Or even worse, some days later I add new tests and they pass, but an uncaught bug is shipped to production.

Reply:
You example above is invalid. Mokito [says](https://github.com/mockito/mockito/wiki/FAQ#mockito-2x-specific-limitations)

> Cannot mock equals(), hashCode(). Firstly, you should not mock those methods. Secondly, Mockito defines and depends upon a specific implementation of these methods. Redefining them might break Mockito.

So this isn't `when/thenReturn`'s fault. :joy: 

Their behaviors are the same for mocked object, and `when/thenReturn` gives better readability and type-checking. For spied object, they behave differently, with `when/thenReturn` calls real method on spied object while `doReturn/when` does not. That's why the convention is to use `doReturn/when` __only when `when/thenReturn` does not work__. :flushed:

Reply:
Well, that's fare enough. I chose a bad method. Let me give you another example, using my own method.

```
  interface A {
    String method();
  }

  static final class B implements A { // same result for non-final
    @Override
    public String method() {
      return "I'm final. Don't stub on me!";
    }
  }

  @Test
  public void whenThenReturnIsAJoke() {
    A a = mock(A.class);
    a.method();

    B b = new B();
    when(b.method()).thenReturn("123");   // stub on b by mistake

    assertThat(a.method()).isEqualTo("123");  // pass!
  }
```

Reply:
LOLOL. I don't know what to say for this case. :sweat: 

We talk about advantages of using one verse the other with the proposition that they are used correctly. 

`P`: "Mockito is used correctly"
`Q`: "`when/thenReturn` is preferred over `doReturn/when`"

We are arguing if `P -> Q` is true. But if `P` is false, then `P -> Q` is true for both `Q` is true and false.

---------------
Just for fun...

Anyway, the style guide mentions this just in favor of `when/thenReturn`'s advantage for readability and type-checking, and that's what we syntactically want.

Reply:
To me, being unsafe or errorprone to use `when/thenReturn` is a bigger concern that outweighs its benefit of slightly better readability and typesafe. `doReturn/when` does not have bad readability as it is already well-known. Yes, `doReturn/when` is not typesafe, but in most cases, if you `doReturn` a value with a wrong type, the test will fail at that line and you are noticed immediately. It's unlike `when/thenReturn`, when it were incorrect the test may not fail at that line, and sometimes the test passes because your main code has a bug and you won't be noticed.
```
when(b.method()).thenReturn("123");   // stub on b by mistake, test passes
```

As for your propositions, how do you check and guarantee that `P` is correct, and `P` continues to be correct forever as your codebase grow? It's impractical to check every existing `when/thenReturn` if they are correct every time you modify/add new code to your test. Why not just use `doReturn/when` from beginning and not to worry about `when/thenReturn` became incorrect later on and need change to `doReturn/when`?